### PR TITLE
Test and reject out-of-bounds shuffle vectors

### LINKED ISSUE #73542: Reject OOB shufflevector intrinsics
Looks like there are some [open](https://blog.regehr.org/archives/1737) questions regarding the semantics of shufflevector intrinsics in LLVM:

> When the mask is out of bounds, the current semantics are that the resulting element is undef. If shufflevector is used to remove an element from the input vector, then the instruction cannot be removed, as the input might have been poison. The solution is to switch to give poison instead. 

So my question is, what are the semantics of those intrinsics in MIR? We had some [discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/LLVM.20IR.20semantics) and it seems like it should be one of the two:
* OOB shufflevector should be statically ruled out. We could even have the new MIR validation pass check that.
* OOB shufflevector returns "uninitialized memory" for the affected elements. (Following [this paper](http://www.cs.utah.edu/~regehr/papers/undef-pldi17.pdf), MIR does not have a poison vs undef distinction; our "uninit" is closest to LLVM's poison.)

The former seems definitely the safest ;) but currently this is [not enforced](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3e5c8da20683871a552d0b4067a3f8cc). @hanna-kruppe was concerned that stdarch might rely on OOB shufflevector somewhere. If we have consensus that for now, we should statically rule out OOB indices to avoid UB/uninit here, I guess the main open issue is (a) implementing that check and (b) fixing the fallout, if any.

Cc @rust-lang/lang @rust-lang/libs @gnzlbg @Lokathor  (not sure who else to ping for stdarch)

### PR DESCRIPTION
**Pull request by workingjubilee**
Fixes #73542.

### CODE REVIEW

**Comment by RalfJung** [comment]
Thanks, this looks good!

Is `simd_shuffle4` the only intrinsic that needs such a check?

**Comment by workingjubilee** [comment]
Well, I can macro the test over the various other LLVM shuffle intrinsics. There are actually a few other tests for this floating in the compiler that verify this error code is emitted (since it basically exists for this issue). As far as I can tell, it would be the shuffles, and maybe insert/extract, that could apply out-of-bounds logic in the first place.

**Comment by RalfJung** [comment]
Yeah, would be good to have all intrinsic bounds checks covered here. Canaries are great but without knowing how the check works internally it is easy to think one has covered all code paths but forget about some odd special case...

When I grep for that error code, most existing tests seem to be about other aspects? In fact I can see none that says anything about "out of bounds". The test should cover all intrinsics that can have that particular error. The error code is for all sorts of problems with intrinsics.

**Comment by workingjubilee** [comment]
That's completely fair, will generalize this fully then (and figure out if insert/extract can go OOB... if they can, I'll include them).

**Comment by RalfJung** [comment]
> That's completely fair, will generalize this fully then (and figure out if insert/extract can go OOB... if they can, I'll include them).

@workingjubilee  From what I can see, this is still outstanding.

**Comment by workingjubilee** [comment]
Indeed it was!

@rustbot modify labels: -S-waiting-on-author, +S-waiting-on-review

**Comment by workingjubilee** [comment]
Force pushes were tidy thrash, whoops.

File: src/test/ui/simd/shuffle-not-out-of-bounds.rs [hunk 1]
```diff
@@ -6,18 +6,185 @@
...
+                    }
+                    let vec1 = $x$t;
+                    let vec2 = $x$t;
+                    $y(vec1, vec2, ARR)
+                };
+        }
     }
 }
+//~^^^^^ ERROR: invalid monomorphization of `simd_shuffle2` intrinsic
+//~^^^^^^ ERROR: invalid monomorphization of `simd_shuffle4` intrinsic
```
Comment (RalfJung) [inline src/test/ui/simd/shuffle-not-out-of-bounds.rs:169]: `//~|` would save some counting of `^` here. ;)

**Comment by RalfJung** [comment]
r=me with or without `//~|`.

File: src/test/ui/simd/shuffle-not-out-of-bounds.rs (hunk 1 above)
Comment (workingjubilee) [reply src/test/ui/simd/shuffle-not-out-of-bounds.rs]: Oh, I just thought I was supposed to get in some archery practice with all these arrows. :sweat_drops:

**Comment by workingjubilee** [comment]
@bors r=RalfJung
does this work? :^)
@rustbot modify labels: -S-waiting-on-author, +S-waiting-on-review if not

**Comment by workingjubilee** [comment]
RIP me. :sweat_drops:

**Comment by RalfJung** [comment]
> RIP me. sweat_drops

Sorry -- next time I'll let bors know that you can do this. :)

**Pull request merged**